### PR TITLE
Hyundai: Car Port for Kona 2022

### PR DESCRIPTION
**Pull request by sunnyhaibin**
**Checklist**
- [x] added entry to CarInfo in selfdrive/car/*/values.py and ran `selfdrive/car/docs.py` to generate new docs
- [x] test route added to [routes.py](https://github.com/commaai/openpilot/blob/master/selfdrive/car/tests/routes.py)
- [x] route with stock system: `3bd70ed1a93771dc|2023-09-19--19-16-45`
- [x] route with openpilot: `3bd70ed1a93771dc|2023-09-23--17-14-54`
- [x] harness Type: [Hyundai O](https://github.com/commaai/openpilot/wiki/Hyundai-Kia-Genesis#harness-guide)
- [x] architecture: CAN

**Link to all physical parts during installation**
  - [comma 3X](https://comma.ai/shop/comma-3x)
  - [car harness kit](https://comma.ai/shop/car-harness)
    - [Hyundai O](https://github.com/commaai/openpilot/wiki/Hyundai-Kia-Genesis#harness-guide)

Thanks to the community 2022 Hyundai Kona owner big_bull_ (Discord).

### CODE REVIEW

**Comment by sshane** [comment]
Can you rebase?

**Comment by sunnyhaibin** [comment]
> Can you rebase?

Done!

**Comment by sshane** [comment]
This user's Kona has a lat accel factor of 3.8, even limited at 270 units. We should understand why it's so high, even with the existing 70% limiting.

**Comment by A2SR** [comment]
I am the user. If there is anything that I can do to help let me know.

**Comment by sshane** [comment]
@A2SR Yes, can you try to test maxing out the torque? This involves taking a turn fast enough (reasonably) that it shows the out of torque warning for a few seconds without touching the wheel. A highway interchange or exit should be a safe place to test.

**Comment by sshane** [comment]
@A2SR any updates?

**Comment by johnnathanwood** [comment]
> @A2SR any updates?

@sshane i can also do this test if needed

**Comment by sshane** [comment]
Sure, if you have the same car!

**Comment by johnnathanwood** [comment]
> Sure, if you have the same car!

Same car. 2022 Kona. Are you able to see my logs? I'll do the test in the am

**Comment by sshane** [comment]
Yeah, I just need a dongle id to check.

**Comment by johnnathanwood** [comment]
> Yeah, I just need a dongle id to check.

0099bdb24d82951b

Check 9:14 am, should have hit the torque limit but didnt. It was a pretty sharp turn

**Comment by sshane** [comment]
Would be preferable to get a little more data if you have time for it! That was a little too low speed and for too little time

**Comment by johnnathanwood** [comment]
> Would be preferable to get a little more data if you have time for it! That was a little too low speed and for too little time. Curvy highway on-ramps are good places to test the max torque at medium-high speeds

Totally! Ill gather some more data this afternoon! 👯‍♂️

**Comment by johnnathanwood** [comment]
> > Would be preferable to get a little more data if you have time for it! That was a little too low speed and for too little time. Curvy highway on-ramps are good places to test the max torque at medium-high speeds
> 
> Totally! Ill gather some more data this afternoon! 👯‍♂️

Okay from 4:30 - 4:50pm

**Comment by johnnathanwood** [comment]
I have been getting “driver grasp not detected. Speed will be limited” randomly when speed > 60mph. Time stamp 8:48 am

**Comment by johnnathanwood** [comment]
Can we rebase?

**Comment by sshane** [comment]
Can you pull and try again? It should not be oversteering as it was.

**Comment by johnnathanwood** [comment]
> Can you pull and try again? It should not be oversteering as it was.

Will do! I'll let you know

**Comment by johnnathanwood** [comment]
> Can you pull and try again? It should not be oversteering as it was.

8:50am-9:00am. Much better imo

**Comment by johnnathanwood** [comment]
Still oversteering slightly. Not as aggressive but still noticeable.

Log: 0099bdb24d82951b|2023-12-20--16-30-42--10

**Comment by johnnathanwood** [comment]
Grasp warning

Route Id: Route Id: 0099bdb24d82951b|2023-12-22--12-01-16--14

**Comment by johnnathanwood** [comment]
@sshane can we also increase the mass value to CAR.KONA_2022: 1509.

**Comment by johnnathanwood** [comment]
##Confirmed

[x] radar/camera scc

**Comment by johnnathanwood** [comment]
> ## Confirmed
> * [x]  radar/camera scc
> 
> @sshane would this be causing the "grasp warning" by chance?

@sshane also this might be the root of the issue too?

**Comment by johnnathanwood** [comment]
@sshane could you take a look at this at some point today to see if we can make it into the 9.6 release?

**Comment by johnnathanwood** [comment]
@adeebshihadeh can we add this to the 9.7 milestone?

**Comment by adeebshihadeh** [comment]
> @adeebshihadeh can we add this to the 9.7 milestone?

I removed it from the milestone since the car team's priorities are currently improving support for existing cars, specifically https://github.com/orgs/commaai/projects/25 and https://github.com/commaai/openpilot/discussions/30805. We'll get back to car ports soon.

**Comment by johnnathanwood** [comment]
@sunnyhaibin can we rebase please

File: selfdrive/car/hyundai/values.py [hunk 1]
```diff
@@ -27,7 +27,7 @@ def __init__(self, CP):
     self.STEER_STEP = 1  # 100 Hz
 
     if CP.carFingerprint in CANFD_CAR:
-      self.STEER_MAX = 270
+      self.STEER_MAX = 170
```
Comment (jyoung8607) [inline selfdrive/car/hyundai/values.py:30]: Looked at this by request of @johnnathanwood. KONA_2022 isn't a CAN_FD car, so this change wasn't effective and he's still testing on full torque. It falls under ALT_LIMITS.

@sshane @sunnyhaibin can one of you push a test update to this branch lowering KONA_2022 to like 190?

Comment (sunnyhaibin) [reply selfdrive/car/hyundai/values.py:30]: > Looked at this by request of @johnnathanwood. KONA_2022 isn't a CAN_FD car, so this change wasn't effective and he's still testing on full torque. It falls under ALT_LIMITS.
> 
> @sshane @sunnyhaibin can one of you push a test update to this branch lowering KONA_2022 to like 190?

@jyoung8607 is this to test the "oversteer" behavior?

**Review by johnnathanwood — COMMENTED** [review]
Handled beautifully and much more useable

log: 0099bdb24d82951b/2024-02-17--10-03-18/9

I have been noticing that has really aggressive micro adjustments and changing lanes feels like it it's cutting over hard.

Log: 0099bdb24d82951b/2024-03-10--16-58-59/29
Log: 0099bdb24d82951b/2024-03-11--08-30-16/10

Here is a torque warning to make sure its not exceeding the limit:

log: 0099bdb24d82951b/2024-03-03--18-43-22/2
log: 0099bdb24d82951b/2024-03-01--19-11-56/2

File: selfdrive/car/tests/routes.py
Comment (johnnathanwood) [inline selfdrive/car/tests/routes.py:1]: 0099bdb24d82951b/2024-02-17--10-03-18/9

File: selfdrive/car/tests/routes.py [hunk 2]
```diff
@@ -139,6 +139,7 @@ class CarTestRoute(NamedTuple):
   CarTestRoute("012c95f06918eca4|2023-01-15--11-19-36", HYUNDAI.IONIQ),  # openpilot longitudinal enabled
   CarTestRoute("ab59fe909f626921|2021-10-18--18-34-28", HYUNDAI.IONIQ_HEV_2022),
   CarTestRoute("22d955b2cd499c22|2020-08-10--19-58-21", HYUNDAI.KONA),
+  CarTestRoute("3bd70ed1a93771dc|2023-09-23--17-14-54", HYUNDAI.KONA_2022),
```
Comment (johnnathanwood) [inline selfdrive/car/tests/routes.py:142]: 0099bdb24d82951b/2024-02-17--10-03-18/9

**Comment by johnnathanwood** [comment]
@sunnyhaibin do we need to block some messages from the cam scc? Still getting the random “grasp warning”. Was not getting it when I tested dev-c3

**Comment by johnnathanwood** [comment]
@adeebshihadeh @sshane can we revisit this?

File: selfdrive/car/hyundai/values.py [hunk 3]
```diff
@@ -43,7 +43,7 @@ def __init__(self, CP):
 
     # these cars have significantly more torque than most HKG; limit to 70% of max
     elif CP.flags & HyundaiFlags.ALT_LIMITS:
-      self.STEER_MAX = 270
+      self.STEER_MAX = 230
```
Comment (sshane) [inline selfdrive/car/hyundai/values.py:46]:
```suggestion
      self.STEER_MAX = 170
```

**Comment by sshane** [comment]
I fixed the torque, please try that out. We won't have time to look into merging this soon though, as we need to detect reliably how your vehicle is different from the other Konas that explains why the steering response is so different.

**Comment by johnnathanwood** [comment]
> I fixed the torque, please try that out. We won't have time to look into merging this soon though, as we need to detect reliably how your vehicle is different from the other Konas that explains why the steering response is so different.



> I fixed the torque, please try that out. We won't have time to look into merging this soon though, as we need to detect reliably how your vehicle is different from the other Konas that explains why the steering response is so different.

@shane I believe the steering response was so different for this because we initially had it under the wrong platform. This car is CAN and it was initially set to CAN FD 

@sunnyhaibin can we rebase this please

**Comment by sshane** [comment]
It still appears that your car achieves more turning lateral acceleration with the same torque command vs. the other Konas which is what we need to figure out and detect.

**Comment by johnnathanwood** [comment]
> It still appears that your car achieves more lateral turning acceleration with the same torque command vs. the other CAN Konas which is what we need to figure out and detect.

Even from this route? [0099bdb24d82951b/2024-04-08--16-55-12]

**Comment by johnnathanwood** [comment]
@sunnyhaibin Can we rebase this please?

**Comment by johnnathanwood** [comment]
@sshane @sunnyhaibin if we set the alt limit in the value, it works to keep the torque at 70% of max.

```
    elif CP.flags & HyundaiFlags.ALT_LIMITS:
      self.STEER_MAX = 190
      self.STEER_DELTA_UP = 2
      self.STEER_DELTA_DOWN = 3
```

```
  HYUNDAI_KONA_2022 = HyundaiPlatformConfig(
    [HyundaiCarDocs("Hyundai Kona 2022", car_parts=CarParts.common([CarHarness.hyundai_o]))],
    CarSpecs(mass=1391, wheelbase=2.6, steerRatio=13.42, tireStiffnessFactor=0.385),
    flags=HyundaiFlags.CAMERA_SCC | HyundaiFlags.ALT_LIMITS,
  )
  ```

**Comment by sshane** [comment]
We've moved the car interfacing code to our [`opendbc`](https://github.com/commaai/opendbc) repository, which is now the new home for car ports and fingerprints. Please re-open your pull request against [`opendbc`](https://github.com/commaai/opendbc) at your convenience by using this command below. This will transform all changes under `selfdrive/car/` to `opendbc_repo/opendbc/car/`. Make sure you have initialized submodules beforehand and have checked out the latest `opendbc` commit.

```bash
PR_NUMBER=33045
curl -L https://github.com/commaai/openpilot/pull/$PR_NUMBER.patch | sed -e 's/selfdrive\/car/opendbc_repo\/opendbc\/car/g' | git apply -v --reject
```

Simply replace the PR number with your own. Once done, add the files, fix any conflicts, and open a new PR. Alternatively, you may start a new PR from scratch if that is easier for you.

**Pull request closed without merging**